src/commands/registration/pasv.js

```diff
@@ -8,15 +8,18 @@ module.exports = {
       return this.reply(502);
     }
 
+    let port;
     this.connector = new PassiveConnector(this);
     return this.connector.setupServer()
     .then((server) => {
-      let address = this.server.options.pasv_url;
+      port = server.address().port;
       // Allow connecting from local
-      if (isLocalIP(this.ip)) {
-        address = this.ip;
-      }
-      const {port} = server.address();
+      if (isLocalIP(this.ip)) return this.ip;
+      let address = this.server.options.pasv_url;
+      if (typeof address === "function") return address(this.ip);
```

Comment:
```suggestion
      if (typeof address === "function") {
        return Promise.resolve(address(this.ip))
          .then(address => ({address, port}))
```

Reply:
Thanks for your patience - here is what I was thinking;

```js
return this.connector.setupServer()
.then((server) => {
  const {port} = server.address();
  let address = this.server.options.pasv_url;
  // Allow connecting from local
  if (isLocalIP(this.ip)) address = this.ip;

  if (typeof address === 'function') {
    return Promise.try(() => address(this.ip)).then((address) => ({address, port}));
  } else {
    return {address, port};
  }
}).then(({address, port}) => {
  const host = address.replace(/\./g, ',');
  const portByte1 = port / 256 | 0;
  const portByte2 = port % 256;
  return this.reply(227, `PASV OK (${host},${portByte1},${portByte2})`);
})
```

Reply:
It is probably best _not_ to assume that the `pasv_url` is a function that returns a promise in this case.